python/cuml/manifold/t_sne.pyx

```diff
@@ -0,0 +1,465 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile = False
+# distutils: language = c++
+# distutils: extra_compile_args = -Ofast
+# cython: embedsignature = True, language_level = 3
+# cython: boundscheck = False, wraparound = False, initializedcheck = False
+
+import cudf
+import cuml
+import ctypes
+import numpy as np
+import inspect
+import pandas as pd
+
+from cuml.common.base import Base
+from cuml.common.handle cimport cumlHandle
+
+from cuml.utils import input_to_dev_array as to_cuda, zeros
+from numba import 
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+from libcpp.memory cimport shared_ptr
+
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+cdef extern from "tsne/tsne.h" namespace "ML" nogil:
+    cdef void TSNE_fit(
+        const cumlHandle &handle,
+        const float *X,
+        float *Y,
+        const int n,
+        const int p,
+        const int dim,
+        int n_neighbors,
+        const float theta,
+        const float epssq,
+        float perplexity,
+        const int perplexity_max_iter,
+        const float perplexity_tol,
+        const float early_exaggeration,
+        const int exaggeration_iter,
+        const float min_gain,
+        const float pre_learning_rate,
+        const float post_learning_rate,
+        const int max_iter,
+        const float min_grad_norm,
+        const float pre_momentum,
+        const float post_momentum,
+        const long long random_state,
+        const bool verbose,
+        const bool intialize_embeddings,
+        bool barnes_hut) except +
+
+
+class TSNE(Base):
+    """
+    TSNE (T-Distributed Stochastic Neighbor Embedding) is an extremely
+    powerful dimensionality reduction technique that aims to maintain
+    local distances between data points. It is extremely robust to whatever
+    dataset you give it, and is used in many areas including cancer research,
+    music analysis and neural network weight visualizations.
+
+    cuML's TSNE implementation handles any # of n_components although specifying
+    n_components = 2 will yield a somewhat extra speedup as it uses the O(nlogn)
+    Barnes Hut algorithm.
+
+    Currently, TSNE only has a fit_transform method. For embedding new data, we
```

Comment:
Docs should describe features that are or are not implemented in the algorithm, but should not be giving promises. Let's remove the `For embedding new data...` and put that in a github issue.